### PR TITLE
Modify Shipping Policies

### PR DESCRIPTION
**Pull request by arjun289**
Modify shipping policy feature as per the new design.
## Motivation and Context

Shipping policy earlier was very basic and the rules were not dynamic. The story now has made the rules very dynamic as it allows each rule to have it's own set of data.

## Describe your changes

- Modifies the shipping rules schema to accommodate adding dynamic
 fields depending on the type of rule identifier.
- The shipping calculator logic is modified to cater for new changes.
- The shipping rules type handle the calculation logic for their respective type.
- Modified the shipping rules seed portion as per the new db design.
- The UI portion for the shipping policy has been re-written.

### Pivotal Story
[delivers #162375179]

------

## Checklist

- [x] I have read [CONTRIBUTING.md][contributing].
- [ ] My code follows the [style guidelines][contributing] of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My code does not generate any (new) [`credo`][credo] and compile-time warnings.
- [ ] I have updated the documentation wherever necessary.
- [ ] I have added tests to cover my changes.

[contributing]: https://github.com/aviabird/snitch/CONTRIBUTING.md
[credo]: https://github.com/rrrene/credo

------
Dear Gatekeeper,

Please make sure that the commits that will be merged or rebased into the base
branch are [formatted according to our standards][commit-style].

> The only additional requirement is that the the PR number must appear in the
> commit title in brackets: `(#XXX)`

[commit-style]: https://github.com/aviabird/snitch/blob/develop/CONTRIBUTING.md#styling

### CODE REVIEW

File: apps/snitch_core/lib/core/data/schema/shipping_rules/order_conditional_free.ex [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  def changeset(%__MODULE__{} = data, params) do
+    data
+    |> cast(params, [:amount])
+  end
+
+  def calculate(package, currency_code, rule, prev_cost) do
+    order = Repo.get(Order, package.order_id) |> Repo.preload(:line_items)
+    total_order_cost = OrderDomain.line_item_total(order)
+
+    min_amount = Money.new!(currency_code, rule.preferences["amount"])
```
Comment (ashish173) [inline apps/snitch_core/lib/core/data/schema/shipping_rules/order_conditional_free.ex]: think of a better name for this; intent is not clear.

File: apps/snitch_core/priv/repo/seed/seeds.exs [hunk 2]
```diff
@@ -61,7 +61,7 @@ Shipping.seed!()
 
 ShippingRules.seed!()
 
-ShippingRules.seed!()
+StockLocation.seed!()
```
Comment (ashish173) [inline apps/snitch_core/priv/repo/seed/seeds.exs]: remove this.

File: apps/snitch_core/test/data/schema/shipping_rules/order_conditional_free_test.exs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  end
+
+  describe "calculate/3" do
+    setup :zones
+    setup :shipping_methods
+    setup :embedded_shipping_methods
+
+    test "order meets the condition and return is {:halt, cost}", context do
+      # if free shipping for order above some amount is applied
+      # it overrides cost set by rule set along with it return is {:halt, cost}
```
Comment (ashish173) [inline apps/snitch_core/test/data/schema/shipping_rules/order_conditional_free_test.exs]: set product price and quantity here. this makes the order value/behaviour explicit.

File: apps/snitch_core/test/data/schema/shipping_rules/order_conditional_free_test.exs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+    test "order does not meet condition and return is {:cont, cost}", context do
+      # if free shipping for order above some amount is not applied
+      # it returns {:cont, cost}
+
+      rule_manifest = %{code: :fsoa, description: "free shipping above amount"}
+      preference_manifest = %{amount: 20}
+
+      %{package: package, rule: rule} =
+        package_with_shipping_rule(context, 1, rule_manifest, preference_manifest)
```
Comment (ashish173) [inline apps/snitch_core/test/data/schema/shipping_rules/order_conditional_free_test.exs]: same as above.

File: apps/snitch_core/test/data/schema/shipping_rules/product_flat_rate_test.exs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    setup :zones
+    setup :shipping_methods
+    setup :embedded_shipping_methods
+
+    test "returns {:cont, cost} for flat cost per product", context do
+      rule_manifest = %{code: :fsrp, description: "fixed shipping rate per product"}
+      preference_manifest = %{cost_per_item: Decimal.new(10.00)}
+
+      %{package: package, rule: rule} =
+        package_with_shipping_rule(context, 3, rule_manifest, preference_manifest)
```
Comment (ashish173) [inline apps/snitch_core/test/data/schema/shipping_rules/product_flat_rate_test.exs]: create a local variable for quantity for this `3`.

File: apps/snitch_core/test/data/schema/shipping_rules/product_flat_rate_test.exs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+      assert {:cont, cost} =
+               ProductFlatRate.calculate(package, currency(), rule, Money.new!(currency(), 0))
+
+      currency = currency()
+
+      assert cost ==
+               currency
+               |> Money.new!(Decimal.new(10.00))
+               |> Money.mult!(3)
```
Comment (ashish173) [inline apps/snitch_core/test/data/schema/shipping_rules/product_flat_rate_test.exs]: use the local variable defined above.

File: apps/snitch_core/test/domain/shipping_calculator_test.exs [hunk 7]
```diff
@@ -4,141 +4,223 @@ defmodule Snitch.Domain.ShippingCalculatorTest do
...
+      rule_active_manifest = %{fso: false, fsoa: true, ofr: true, fsrp: false}
+
+      preference_manifest = %{
+        fso: %{},
+        fsoa: %{amount: 50},
+        ofr: %{cost: 20},
+        fsrp: %{cost_per_item: 5}
+      }
+
+      %{package: package, category: category} = setup_package_with_sc(context, _quantity = 7)
```
Comment (ashish173) [inline apps/snitch_core/test/domain/shipping_calculator_test.exs]: make product price explicit here.

File: apps/snitch_core/test/domain/shipping_calculator_test.exs [hunk 8]
```diff
@@ -4,141 +4,223 @@ defmodule Snitch.Domain.ShippingCalculatorTest do
...
-      # order rule is set to false so it won't override product rule
+    test "with free for order above amount :fsoa does not apply", context do
+      rule_active_manifest = %{fso: false, fsoa: true, ofr: true, fsrp: false}
 
-      rule_active_manifest = [true, false, true, false]
-      cost_manifest = [Money.new!(:USD, 100), Money.new!(:USD, 100), Money.new!(:USD, 10)]
+      preference_manifest = %{
+        fso: %{},
+        fsoa: %{amount: 200},
+        ofr: %{cost: 20},
```
Comment (ashish173) [inline apps/snitch_core/test/domain/shipping_calculator_test.exs]: have a local variable for `20`.

File: apps/snitch_core/test/domain/shipping_calculator_test.exs [hunk 9]
```diff
@@ -4,141 +4,223 @@ defmodule Snitch.Domain.ShippingCalculatorTest do
...
+      setup_shipping_rules(rule_active_manifest, preference_manifest, category)
 
       shipping_cost = ShippingCalculator.calculate(package)
 
-      # since total order cost would be less fixed rate for order is
-      # applied
-      assert shipping_cost == Money.new!(:USD, 10) |> Money.round()
+      # as free shipping is available over 200, fixed rate cost is applied
+      # unit_price 9.99 * 1 < 200
+      assert shipping_cost == Money.new!(currency(), 20) |> Money.round()
```
Comment (ashish173) [inline apps/snitch_core/test/domain/shipping_calculator_test.exs]: use here the local variable.

File: apps/snitch_core/test/data/schema/shipping_rules/order_conditional_free_test.exs (hunk 3 above)
Comment (arjun289) [reply apps/snitch_core/test/data/schema/shipping_rules/order_conditional_free_test.exs]: have made the suggested change to all the documents having this behaviour to keep things explicit.

File: apps/snitch_core/test/domain/shipping_calculator_test.exs (hunk 7 above)
Comment (arjun289) [reply apps/snitch_core/test/domain/shipping_calculator_test.exs]: made all the changes suggested for this file.

**Pull request merged**